llvm-external-projects/iree-dialects/include/iree-dialects/Transforms/TransformMatchers.h

```diff
@@ -135,12 +153,28 @@ class StructuredOpMatcher {
   /// Matches the given operation, hook for `matchPattern`.
   bool match(Operation *op);
 
+  //===-------------------------------------------------------------------===//
+  // Constraints on op rank and dims.
+  //===-------------------------------------------------------------------===//
+  /// Adds a predicate checking that the given rank must be greater than some
+  /// constant value.
+  // TODO: Base class, derived class and proper API.
+  StructuredOpMatcher &rank(NumGreaterEqualTo minRank);
+  StructuredOpMatcher &rank(NumLowerEqualTo maxRank);
+
   /// Adds a predicate checking that the given iteration space dimension is
   /// static/dynamic. The dimension index may be negative, in which case
   /// dimensions are counted from the last one (i.e. Python-style), or be an
   /// AllDims tag, in which case all dimensions are checked. This may be
   /// eventually extended to slices and/or lists of dimensions.
-  StructuredOpMatcher &dim(int64_t dimension, ShapeKind kind);
+  /// When strictInBounds is false, matching succeeds if the dimension is out of
+  /// bounds.
+  StructuredOpMatcher &dim(int64_t dimension, ShapeKind kind,
+                           bool strictInBounds = true) {
```

Comment:
Let's have a tag struct instead of a bool here. Just declare `struct StrinctInBounds` like we do for predicates.